controllers/apim/ratelimitpolicy_controller.go

```diff
@@ -136,100 +145,113 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
 			return err
 		}
-		logger.V(1).Info("successfully added ratelimitpolicy annotation to virtualservice")
+		logger.V(1).Info("successfully added RateLimitPolicy annotation to VirtualService")
 	}
 
 	// TODO(rahulanand16nov): store context of virtualservice in RLP's status block and manage envoy patches.
-	if err := r.reconcileRateLimitFilters(ctx, vs, rlp); err != nil {
-		logger.Error(err, "failed to reconcile ratelimit filters")
-		return err
-	}
 
-	if err := r.reconcileRateLimitActions(ctx, vs, rlp); err != nil {
-		return err
-	}
-	logger.Info("successfully reconciled ratelimitpolicy using attached virtualservice")
-	return nil
-}
+	// create/update EnvoyFilter patches for each gateway
+	for _, gw := range vs.Spec.Gateways {
+		gwKey := common.NamespacedNameToObjectKey(gw, vs.Namespace) // Istio defaults to VirtualService's namespace
+		gwLabels := gatewayLabels(ctx, r.Client(), gwKey)
+		owner := rlpKey.String()
+
+		// create/update ratelimit filters patch
+		// fetch already existing filters patch or create a new one
+		filtersPatckKey := client.ObjectKey{Namespace: gwKey.Namespace, Name: rlFiltersPatchName(gwKey.Name)}
+		filtersPatch := &istio.EnvoyFilter{}
+		if err := r.Client().Get(ctx, filtersPatckKey, filtersPatch); err != nil {
+			if !apierrors.IsNotFound(err) {
+				logger.Error(err, "failed to get ratelimit filters patch")
+				return err
+			}
+			filtersPatch, err = ratelimitFiltersPatch(gwKey, gwLabels)
+			if err != nil {
+				logger.Error(err, "failed to form ratelimit filters patch")
+				return err
+			}
+		}
 
-func (r *RateLimitPolicyReconciler) reconcileRateLimitActions(ctx context.Context, vs *istio.VirtualService, rlp *apimv1alpha1.RateLimitPolicy) error {
-	logger := r.Logger()
+		// make sure annotation map is initialized
+		filtersPatchOwnerList := []string{}
+		if filtersPatch.Annotations == nil {
+			filtersPatch.Annotations = make(map[string]string)
+		}
 
-	// TODO(rahulanand16nov): right now patch only applies to kuadrant-gateway but ideally should choose from virtualservice.
-	for _, host := range vs.Spec.Hosts {
-		for _, route := range rlp.Spec.Routes {
-			isRouteMatch := doesRouteMatch(vs.Spec.Http, route.Name)
-			if !isRouteMatch {
-				logger.V(1).Info("no match found for route", "route name", route.Name)
-				continue
-			}
+		if ogOwnerRlps, present := filtersPatch.Annotations[envoyFilterAnnotationOwnerRLPs]; present {
+			filtersPatchOwnerList = strings.Split(ogOwnerRlps, ownerRlpSeparator)
+		}
 
-			vHostName := host + ":80" // Istio names virtual host in this format: host + port
-			routePatch := routeRateLimitsPatch(vHostName, route.Name, route.Actions, route.Stage)
+		// add the owner name if not present already
+		if !common.Contains(filtersPatchOwnerList, owner) {
+			filtersPatchOwnerList = append(filtersPatchOwnerList, owner)
+		}
+		finalOwnerVal := strings.Join(filtersPatchOwnerList, ownerRlpSeparator)
```

Comment:
Have you considered the option of having one annotation per owner in the EnvoyFilter object?

```ỳaml
metadata:
  annotations:
     kuadrant.io/ownerRateLimitPolicies-rlp1: true
     kuadrant.io/ownerRateLimitPolicies-rlp2: true
     kuadrant.io/ownerRateLimitPolicies-rlp3: true
```

The value of the annotations is meaningless, but it feels easier to:
* read for a human
* easier to manage from the code. The operator to check if the owner already exists is straightforward.